src/merkle_tree.rs

```diff
@@ -30,155 +47,142 @@ impl MerkleNode {
-                added_proof = true;
-            }
-            if Self::recursive_get_proof(left, proof, transaction_hash) {
-                current_node.right_son.as_ref().inspect(|right_sibling| {
-                    proof.push(SiblingHash::Right(right_sibling.hash_value))
-                });
-                added_proof = true;
-            }
-        });
-
-        current_node.right_son.as_ref().inspect(|right| {
-            if right.hash_value == transaction_hash {
-                current_node
-                    .left_son
-                    .as_ref()
-                    .inspect(|left_sibling| proof.push(SiblingHash::Left(left_sibling.hash_value)));
-                added_proof = true;
-            }
-            if Self::recursive_get_proof(right, proof, transaction_hash) {
-                current_node
-                    .right_son
-                    .as_ref()
-                    .inspect(|left_sibling| proof.push(SiblingHash::Left(left_sibling.hash_value)));
-                added_proof = true;
-            }
-        });
+        match current_node {
+            MerkleNode::Inner(node) => {
+                if node.left_son.get_hash_value() == transaction_hash {
+                    proof.push(SiblingHash::Right(node.right_son.get_hash_value()));
+                    return true;
+                }
+                if Self::recursive_get_proof(&node.left_son, proof, transaction_hash.clone()) {
+                    proof.push(SiblingHash::Right(node.right_son.get_hash_value()));
+                    return true;
+                }
 
-        added_proof
+                if node.right_son.get_hash_value() == transaction_hash {
+                    proof.push(SiblingHash::Left(node.left_son.get_hash_value()));
+                    return true;
+                }
+                if Self::recursive_get_proof(&node.right_son, proof, transaction_hash) {
+                    proof.push(SiblingHash::Left(node.left_son.get_hash_value()));
+                    return true;
+                }
+
+                false
+            }
+            MerkleNode::Leaf(_) => false,
+        }
     }
 
     pub fn get_proof(&mut self, transaction: H) -> Vec<SiblingHash> {
         let mut proof = Vec::new();
-        let mut hasher = DefaultHasher::new();
-        transaction.hash(&mut hasher);
-        Self::recursive_get_proof(&self.merkle_root, &mut proof, hasher.finish());
+        let mut hasher = Sha3_256::new();
+        hasher.update(transaction);
+        Self::recursive_get_proof(
+            &self.merkle_root,
+            &mut proof,
+            format!("{:X}", hasher.finalize()),
+        );
         proof
     }
 
     pub fn add(&mut self, transaction: H) -> Result<(), &'static str> {
-        self.leafs.push(transaction);
-        self.merkle_root = Self::create_tree(self.leafs.clone())?.merkle_root;
+        self.leaves.push(transaction);
+        self.merkle_root = Self::create_tree(self.leaves.clone())?.merkle_root;
```

Comment:
Two things:
1. It's not necessary to recreate the whole tree to add a leaf. You'll have to update the hashes of one path from the root to the new leaf, but otherwise the rest should remain valid.
2. Even then, cloning them is not necessary: you can instead replace the whole tree, so you can pass ownership. E.g.:
```rust
let leaves = self.leaves; // may need `std::mem::take(&mut self.leaves)` instead, if the borrow checker doesn't see it's safe to move
*self = Self::create(leaves)?;
```
That way you don't have to deep copy and then drop the previous copy.

Reply:
Going to read about the `std::mem` package, didn't really know about it